### ISSUE TITLE
Tables without rowid causes "no such column"-logs

### ISSUE THREAD

**Issue by simme** [groue/GRDB.swift#945]
### What did you do?

Created a table without rowid.
```swift
db.create(table: "product_categories", withoutRowID: true)
```

### What did you expect to happen?

Nothing special really.

### What happened instead?

Everytime I access the database through my `FetchableRecord`s I get these log entries.
`no such column: rowid in "SELECT rowid FROM "list_items""`

The reason my tabels are without rowid is because I use UUIDs to identify everything. Perhaps this is the wrong approach?

### Environment

**GRDB flavor(s):** GRDB
**GRDB version:**  5.6.0
**Installation method:** SPM
**Xcode version:** 12.3
**Swift version:** 5.3
**Platform(s) running GRDB:** iOS
**macOS version running Xcode:** 11.2.3

### Demo Project

If necessary for further debugging I can certainly provide this! :)

**Comment by groue** [comment]
Hello @simme,

I know GRDB has some blind spots about "without rowid" tables, but it is not intended that "everytime [one] accesses the database" an error happens.

Can you be more specific (and tell precisely about errors, instead of logging errors away)?

**Comment by simme** [comment]
Hey!

Sorry if I was unclear. There's no error per se. Just the log statement. Everything seems to work fine.
I'll see if I can narrow down the cause a bit. But I'm pretty sure it's related to starting a `ValueObservation`. Like so:

```swift
    ValueObservation
      .tracking { db -> [ProductCategoryModel] in
        try ProductCategoryModel.fetchAll(db)
      }
      .publisher(in: pantry.databasePool).map {
        $0.map { ProductCategory(model: $0, locale: .current) }
      }
      .replaceError(with: [])
      .eraseToAnyPublisher()
```

**Comment by simme** [comment]
I'm in the middle of the last 10% of the app I'm working on at the moment, so not a ton of extra time. Since this only leads to the log message and everything seems to work there's no panic. Maybe everything is working as intended, but I'm holding it wrong :)

**Comment by groue** [comment]
OK. Do you happen to reproduce with a simple: 

```
try pantry.databasePool.read { db in
    try ProductCategoryModel.fetchAll(db)
}
```

What I ask here is your help building a *minimal* reproducing case, if you have time.

**Comment by simme** [comment]
That does not cause the log statement.

Sorry for the overly inclusive statement in the initial ticket. I do all my fetching through `ValueObservation`s so I was sort of right that it's `Everytime I access my database`.

I'm stepping through the code now trying to see where it originates!

**Comment by groue** [comment]
I'm trying to reproduce, with no success. I have to ask for a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example).

Also, mind that database observation can't see changes to WITHOUT ROWID tables. This information is little bit [buried](https://github.com/groue/GRDB.swift#database-changes-and-transactions) in the GRDB doc (I'll make it more prominent). It's a [known SQLite limitation](https://www.sqlite.org/withoutrowid.html):

> The sqlite3_update_hook() interface does not fire callbacks for changes to a WITHOUT ROWID table

**Comment by simme** [comment]
Yup, totally understand! Thank you for your time.

> Also, mind that database observation can't see changes to WITHOUT ROWID tables.

Oh shoot. I had completely missed that. And it does explain a lot! I haven't yet dove into why my observations doesn't trigger. Currently all of the changes ot a user's data is reflected in an in-memory model as well. But I was going to notice this once I get to implementing sync.

If this issue only happens in `ValueObservation`s on WITHOUT ROWID tables anyway (that's only where I see it) I guess the whole ticket is moot. Does GRDB know if an observed table is WITHOUT ROWID and could it warn when trying to observe one? Just a thought.

**Comment by groue** [comment]
> If this issue only happens in `ValueObservation`s on WITHOUT ROWID tables anyway (that's only where I see it) I guess the whole ticket is moot.

I'm afraid I can't ask you solving an issue which is not yours, indeed :-)

> Does GRDB know if an observed table is WITHOUT ROWID and could it warn when trying to observe one? Just a thought.

This is a good idea. We already have [PrimaryKeyInfo.tableHasRowID](https://github.com/groue/GRDB.swift/blob/v5.6.0/GRDB/Core/Database%2BSchema.swift#L662) which tells if a table is WITHOUT ROWID or not. So I guess we could use this information when ValueObservation [learns about its tracked region](https://github.com/groue/GRDB.swift/blob/v5.6.0/GRDB/ValueObservation/ValueObserver.swift#L220). I'll favorably consider a pull request.

**Comment by simme** [comment]
What would the desired behavior be? As a user of the framework I would probably not hate if GRDB threw an error when trying to create such an observation. Since it technically _can't_ observe properly, and any such usages should most likely be replaced with a regular fetch instead. But I have only surface level knowledge and am certainly ignoring a lot of use cases.

**Comment by groue** [comment]
BTW, I found the origin of your log: https://github.com/groue/GRDB.swift/blob/v5.6.0/GRDB/Core/Database%2BSchema.swift#L189-L198

GRDB is wondering if one table is WITHOUT ROWID or not, and has no other way to find out than to check if selecting `rowid` gives an error or not.

SQLite logs all errors when [you ask for it](https://github.com/groue/GRDB.swift/blob/master/README.md#error-log) through GRDB, but also when some third-party tool registers a global SQLite log handler.

**Comment by groue** [comment]
> What would the desired behavior be? As a user of the framework I would probably not hate if GRDB threw an error when trying to create such an observation. Since it technically _can't_ observe properly, and any such usages should most likely be replaced with a regular fetch instead. But I have only surface level knowledge and am certainly ignoring a lot of use cases.

This looks like the definition of a programmer error (can't be handled at runtime, requires code change), and those deserve a fatal error, don't you think?

**Comment by groue** [comment]
Flagging "needs revisiting", since this issue has revealed:

- that checking for WITHOUT ROWID tables creates undesired logs (maybe we can avoid that)
- that ValueObservation should fatalError when asked to track a WITHOUT ROWID table
- the documentation for database observation should warn much more prominently against WITHOUT ROWID tables

**Comment by simme** [comment]
> BTW, I found the origin of your log: https://github.com/groue/GRDB.swift/blob/v5.6.0/GRDB/Core/Database%2BSchema.swift#L189-L198
> GRDB is wondering if one table is WITHOUT ROWID or not, and has no other way to find out than to check if selecting rowid gives an error or not.

Confirmed with my app that that's where it happens!

> This looks like the definition of a programmer error (can't be handled at runtime, requires code change), and those deserve a fatal error, don't you think?
I was thinking that first, but felt too extreme to suggest 😅 But I completely agree.

**Comment by simme** [comment]
Looking at a potential PR.

Would enumerating the [`observedRegion`](https://github.com/groue/GRDB.swift/blob/v5.6.0/GRDB/ValueObservation/ValueObserver.swift#L220) of the `ValueObserver` and checking if the observed table regions' `rowId` property is `nil` be sufficient?

**Comment by groue** [comment]
> > This looks like the definition of a programmer error (can't be handled at runtime, requires code change), and those deserve a fatal error, don't you think?
> 
> I was thinking that first, but felt too extreme to suggest 😅 But I completely agree.

No quarter! 🤘

> Would enumerating the [`observedRegion`](https://github.com/groue/GRDB.swift/blob/v5.6.0/GRDB/ValueObservation/ValueObserver.swift#L220) of the `ValueObserver` and checking if the observed table regions' `rowId` property is `nil` be sufficient?

Excuse me, I'm not sure I follow?

**Comment by simme** [comment]
Answering my own question. A `nil` `rowids` property on `TableRegion` does not necessarily mean that there are no rowids:

From *DatabaseRegion:262*
```swift
var rowIds: Set<Int64>? // nil means "all rowids"
```

So can't use that to determine if any of the observed tables lack `rowid` and thus fatal error.

**Comment by groue** [comment]
Yes :-)

When the observedRegion is acquired, we have a database connection (`db`). We can iterate the tables in the region, ask the database for the primary key, and ask the primary key if there is a row id:

```swift
let pkInfo = try db.primaryKey(table)
if !pkInfo.tableHasRowID { boom: not observable }
```

**Comment by groue** [comment]
The `rowIds` in DatabaseRegion helps defining a region made of specific rows. For example:

```swift
// Only triggered when player 1 is updated/inserted/deleted
ValueObservation.tracking { db in try Player.fetchOne(db, key: 1) }
```

**Comment by groue** [comment]
Next GRDB release will check for WITHOUT ROWID tables with a distinctive query:

```sql
SELECT rowid AS checkWithoutRowidOptimization FROM player
```

This will help future investigations about the presence of this query in the error log.

I did not find, so far, a way to know if a table is WITHOUT ROWID or not without checking if a query fails to compile or not. Maybe we'll find a better way in the future.

**Comment by groue** [comment]
GRDB 6 beta 4 (about to be shipped) will detect if a table was created with the `WITHOUT ROWID` optimization with [PRAGMA table_list](https://www.sqlite.org/pragma.html#pragma_table_list), when available (SQLite 3.37+, iOS 15.4+, macOS 12.4+, tvOS 15.4+, watchOS 8.5+, custom SQLite builds or SQLCipher).

This will avoid compiling a faulty statement, and polluting the error log.
